### ISSUE TITLE
Support deploying dynamic route to static host

### ISSUE THREAD

**Issue by brillout** [vikejs/vike#1476]
### Description

> I want to handle routes like /todos/@todoId. I’m looking to keep absolutely everything client side but still have the dynamic routes.

This quotes a user who wants to deploy to a static host.

Conversation & design how we can support that: https://github.com/vikejs/vike/discussions/1475.

**Comment by nitedani** [comment]
> Hi yall, I'm looking into using vike with react, and I was wondering if this use case was possible... I want to have some pages as SPA and others as SSG, and deploy to static hosting like Netlify.
Now this is in general very simple, I just prerender every page, including all SPA pages. But here's the twist, would it be possible to generate only one index file for all the SPA pages and use a Netlify config to redirect all of those pages to that file? This is exactly what you'd do with a normal Vite setup, but I just want to add some SSG files to the mix

**Comment by brillout** [comment]
I'm de-prioritizing this. While I personally think that this is a great feature, I still think the other `v1-release` tickets to be more important. That said, I'm happy to revisit prioritization if it's a blocker for someone and if no other framework supports this. I really wonder: is there any framework out there that supports this?

**Comment by andriygm** [comment]
this would be huge -- I know at least [solid-router](https://github.com/solidjs/solid-router) is able to do this. currently cannot use Vike on one of my Solid projects because of this.

**Comment by brillout** [comment]
> this would be huge

Agreed.

> [solid-router](https://github.com/solidjs/solid-router) is able to do this

That isn't an SSR/SSG framework. It's definitely possible to implement, but it isn't trivial either. Thus I'm genuinely curious if there is an SSR/SSG framework out there that supports this.

> currently cannot use Vike on one of my Solid projects because of this

I've added a workaround to my original post. Would that work for you?

**Comment by andriygm** [comment]
perf, I'll head that route (get it) until Vike's router gets support for this 👌

**Comment by CookedApps** [comment]
We'd appreciate this feature as well. 🙌

Is there an example of how to set up the workaround with react-router?

**Comment by ConnorLanglois** [comment]
> That isn't an SSR/SSG framework. It's definitely possible to implement, but it isn't trivial either. Thus I'm genuinely curious if there is an SSR/SSG framework out there that supports this.

If I understand this issue correctly, I believe NextJS supports this kind of routing. Note the new app router in nextjs does not yet support this, but [they are working on it](https://github.com/vercel/next.js/issues/54393#issuecomment-1871629201). Notably, the "old" pages router supports this as other users mention; although I haven't used that, I'm guessing they're referring to the use of code like this:

```js
const router = useRouter()
return <p>Post: {router.query.slug}</p>
```

from [this page](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes).

I also just noticed another user in that issue thread [said](https://github.com/vercel/next.js/issues/54393#issuecomment-1710657144) that Gatsby also supports dynamic routers in static pages.

**Comment by brillout** [comment]
Hm, I'm not sure. From quickly glazing over the links you posted, it still doesn't seem to me that Next.js supports this (both App Router and Pages Router).

An easy way to test is this: can you create a Next.js app that defines a parameterized route but outputs only one `.html` file (for multiple URLs)?

> Gatsby allows you to use dynamic routers in static exports without configuring server-side redirects.

Vike can do that as well. The question is whether you always need to provide a list of URLs for parameterized routes.

Same test: can you create a Gatsby app that defines a parameterized route but outputs only one `.html` file (for multiple URLs)?

**Comment by ConnorLanglois** [comment]
> An easy way to test is this: can you create a Next.js app that defines a parameterized route but outputs only one `.html` file (for multiple URLs)?

Yes I believe so. [Here is a test repo that demonstrates this](https://github.com/ConnorLanglois/vike-1476) (you can look at the `pages/product` folder for the dynamic route, the other stuff is nextjs scaffolding defaults etc). When you build it, it will output a single html file for the parameterized route (`[id].html`) and the routing is done client side. Let me know if this differs from what you were asking, seems like "dynamic routes" is an overloaded term so trying to make sure we match up on what we mean.

For Gatbsy, I haven't used it before so not sure.

**Comment by andriygm** [comment]
>But if you are already on the index page and click the Link to the dynamic route, it does work.
If you reroute a request from say `/product/54387` to `[id].html`, does `[id].html` route on load? or is client routing only done on navigation from within?

**Comment by ConnorLanglois** [comment]
Hello, by reroute do you mean if I go to `/product/54387` directly in the address bar and have the server rewrite it to the `[id].html` file does it load that `[id].html` file?

**Comment by andriygm** [comment]
>but cloudflare and others do or your own express server can

Sure, but the question here (I think) is whether or not the pre-rendered dynamic route page (in this case `[id].html`) routes to say `/product/54387` when you serve `[id].html`.

**Comment by ConnorLanglois** [comment]
Yeah it will still show `/product/54387` in the address bar when cloudflare serves `[id].html`

**Comment by brillout** [comment]
I'm labeling this ticket as `high-prio` and `v1-release`. The fact that Next.js's Pages Router (and soon also its App Router) support this makes this a Vike-specific blocker.

Thanks @ConnorLanglois for the example.

**Comment by brillout** [comment]
As always: feel free to add a comment here if this a blocker specifically for your app.

**Comment by Taujor** [comment]
Personally I have switched from vike to nuxt (which is a bit overkill for my project) due to this feature lacking. I would really love to see this come to vike.

**Comment by TimJohns** [comment]
I'm not sure if the following is a GOOD approach, but it's working for us so far, so wanted to share it, and also ask whether it's supported or not. We use Vue, Vike, and vike-vue. The site is hosted as a static site from a Google Cloud bucket, uses SSG pages in a hybrid (SPA-like) fashion, and gets all runtime data from a separate API. 

We want both client-side routing and deep linking, while still hosting from a static site. For instance, users will often want to deep link to their custom road trips (e.g. https://epicroadtripplanner.com/roadtrips/d9cc2d68-09d0-4d9e-bd0a-258936c2be65), but the fetched content (and the route itself) for roadtrips are dynamic. 

We use file system routing in this project, and all of the following are in the file system at /src/pages/roadtrips/@roadtripId:

"+onBeforePrerenderStart.ts" pre-renders an "all" page:
```
export async function onBeforePrerenderStart() {
  return ["/roadtrips/all"]
}
```

Importantly, we use Google Cloud Load Balancer's url map functionality to re-write the incoming request urls, so that it serves this 'all/index.html' page for any requests to /roadtrips/{roadtripid=*}, regardless of the value of roadtripId in the actual request:

```
pathMatchers:
- defaultService: <redacted>
  name: static
  routeRules:
  - matchRules:
    - pathTemplateMatch: /roadtrips/{roadtripid=*}
    priority: 2100
    routeAction:
      urlRewrite:
        pathTemplateRewrite: /roadtrips/all/index.html
```

Here's the part I'm not sure about - specifically, whether is supported/documented to use resolveRoute, especially in the data hook, outside of a route function:

In the client-only data hook "+data.client.ts", "pageContext.routeParams" still returns "all" for pageContext.routeParams["roadtripId"], so instead we use `resolveRoute` there, to (re)parse urlPathname:
```
export async function data(pageContext: PageContext) {
  const { routeParams } = resolveRoute('/roadtrips/@roadtripId', pageContext.urlPathname);
  const roadtripId = routeParams["roadtripId"];

// <snip> Call a separate backend API w/roadtripId, to get the user's roadtrip data, and return the roadtrip so that it is rendered in +Page.vue
...
```

**Comment by brillout** [comment]
@TimJohns Yes LGTM, I don't see any issues with your appraoch. As for `pageContext.routeParams` it's because it comes from the "server-side" (which is pre-rendered thus resolved to `all`) &mdash; what you're doing makes sense. Btw. would your company be up for [sponsoring](https://github.com/vikejs/vike/issues/1350)? It would be a lovely excuse for implementing this feature before the [`highest-priority` tickets](https://github.com/vikejs/vike/issues?q=is%3Aissue+is%3Aopen+label%3A%22highest-prio+%F0%9F%94%A5%22) :slightly_smiling_face:

**Comment by niutech** [comment]
There is also another approach to dynamic URLs, at least on Github Pages: using 404.html redirection. When users go to e.g. `/todos/{todoID}`, the HTML page is not found, so the 404.html is loaded instead with the following `<script>`:
```
var todoMatch = location.href.match(/todos\/(.+)$/;
if(todoMatch) location.href = '/todos/?' + todoMatch[1];
```
Then in todos/index.html you can refer to todoId using `location.search.substring(1)`.

**Comment by itsezc** [comment]
I'm also blocked by this, I understand that theres a workaround but to make it easier is there a "working example" with the workaround? 

In the long term, I wish there is a way to approach this without React Router, feels like I have two routers then? (Vike Router & React Router)

**Comment by CookedApps** [comment]
@itsezc 
We worked around this by using a simple hash routing approach. This doesn't require a third party router and works with Vike router as well (i.e. doesn't break it). 

In our case, we just needed a way to have a user's profile under a dynamic route. Instead of having routes like `/profile/username` we just have `/profile#username`. We can read this "path" simply by using `window.location.hash`. We wrote some React hooks to make our lives easier. 

This approach is fairly limited and will not fit every use case. Thus, I agree that **dynamic routes are a necessary feature that Vike Router needs**. And writing "Vike's built-in router has many features that React Router doesn't (and cannot) offer." in the docs is misleading to say the least.

**Comment by brillout** [comment]
@itsezc AFAICT you won't need React Router after this ticket is implemented.

@CookedApps c7ce3a1086846cf2a1a194b028f20af8f0a447d5 and feel free to elaborate what React Router features you're missing when using Vike's router.

**Comment by CookedApps** [comment]
@brillout 
> feel free to elaborate what React Router features you're missing when using Vike's router.

Client-only dynamic routes.

**Comment by itsezc** [comment]
> @itsezc AFAICT you won't need React Router after this ticket is implemented.
> 
> @CookedApps [c7ce3a1](https://github.com/vikejs/vike/commit/c7ce3a1086846cf2a1a194b028f20af8f0a447d5) and feel free to elaborate what React Router features you're missing when using Vike's router.

Good to hear @brillout, this is a huge blocker for me while use Vike for personal projects (later professionally), would it be possible to have a working example with the "workaround" which could be used as a reference (with React Router or another package you would suggest) and what sort of ETA are we looking at for this? I'm aware that its listed for v1, but just trying to "plan" ahead - won't hold you to anything.

**Comment by brillout** [comment]
Bumping to `highest-prio`. I'll work on this ticket after I'm done with Vike's CLI and `vike.config.js`. Indeed, there isn't any ETA but I think I'll be done with it within this year.

**Comment by anuraaga** [comment]
I ran into this as well using Firebase hosting, which allows rewrites for static generated SPAs - while I think the question about other frameworks was answered, FWIW I have had this work with other frameworks, NextJS as mentioned, and many many years ago with a webpack ssg plugin of some sort.

I found an acceptable workaround for me, don't think I see it so posting in case it helps. It just defines a fallback page that always nagivates to the URL, and rewrites to it, not index.html as is otherwise common.

```ts
import { useEffect } from "react";
import { usePageContext } from "vike-react/usePageContext";
import { navigate } from "vike/client/router";

export default function Page() {
  const pageContext = usePageContext();
  useEffect(() => {
    navigate(pageContext.urlOriginal);
  }, [pageContext]);
  return undefined;
}
```

Firebase config, most static hosts more sophisticated than the most basic would have a similar setting

```json
{
  "hosting": {
     ...
     {
        "source": "**",
        "destination": "/fallback/index.html"
      }
  }
}
```

I guess something built-in to Vike could behave a bit better, in my case the navbar I have in my Layout renders before the page while ideally nothing is rendered before invoking the correct route. A further workaround would be to separate the `Layout` for this page, but without #1692 it means having two folders, one for the app and one for just this fallback which is not worth it for my current use case since I can wait for the proper fix in this issue.

Thanks for the great framework, while this isn't the only edge I've run into, they're all minor and I'm having a good time.

**Comment by brillout** [comment]
Thank you for sharing your workaround.

>  this isn't the only edge

Feel free to elaborate in a new GitHub discussion.

**Comment by tbscode** [comment]
Currently I'm working on a golang backend that used vike as static site generator. ( but also allows vike client side navigation ).

Give the default vike example what I do is:

```bash
export async function onBeforePrerenderStart() {
  return [`star-wars`, `star-wars/{id}`]
}
```

Then from my backend I can automatically route any path that matches `star-wars/{id}` to `star-wars/{id}/index.html`.

BUT that also means that for all the `star-wars/{id}/index.html` I need to manually replace `pageContext.routeParams.id` with the correct value, as currently it would be hard-coded to `{id}` on the `star-wars/{id}/index.html` page, thus currently I can just do `html.replace("{id}", path value extracted by backend )`.

What I would find incredibly useful if e.g.: one could just add the params inside the route inside `onBeforePrerenderStart` so there `/star-wars/@id` if then vike could be smart enough to determine the `@id` parameter automatically in the client and resolve it correctly on any `star-wars/@id` page. 

e.g.: for the page `/star-wars/1` one would render the document `/star-wars/@id/index.html` but the vike client would still automatically set `pageContext.routeParams.id = 1` ( and not `pageContext.routeParams.id  = @id` as it would currently ).

Possibly this could also be the default SSG behavior if no `onBeforePrerenderStart` is specified.

**Comment by brillout** [comment]
That's the plan, although a warning will still be shown (that you can suppress) because the pre-rendered HTML won't (cannot) include the content of the page.

**Comment by brillout** [comment]
How about the `onHydrationEnd()` workaround described at https://vike.dev/pageContext#pre-rendering?

TLDR:
 - Use `/product?id=42` instead of `/product/42`.
 - Define `/product/+Page.js` that renders an empty shell (instead of defining `/product/@id/+Page.js`).
 - Re-render the whole page in `onHydrationEnd()` in order to replace the empty shell with the full content.

This should already work today.

One neat thing about this approach is that it doesn't require URL rewriting: I guess static hosts also serve search params?

Maybe there could some kind of Vike option for prettifying the URL to have `/product/42` instead of `/product?42`  🤔 Not sure how to achieve this without a convoluted/confusing DX.

**Comment by itsezc** [comment]
> How about the `onHydrationEnd()` workaround described at https://vike.dev/pageContext#pre-rendering?
> 
> TLDR:
> 
>     * Use `/product?id=42` instead of `/product/42`.
> 
>     * Define `/product/+Page.js` that renders an empty shell (instead of defining `/product/@id/+Page.js`).
> 
>     * Re-render the whole page in `onHydrationEnd()` in order to replace the empty shell with the full content.
> 
> 
> This should already work today.
> 
> One neat thing about this approach is that it doesn't require URL rewriting: I guess static hosts also serve search params?
> 
> Maybe there could some kind of Vike option for prettifying the URL to have `/product/42` instead of `/product?id=42` 🤔 Not sure how to achieve this without a convoluted/confusing DX.

This does work (in theory) but what about the implications of SEO? Search engines generally prefer full URLs

**Comment by brillout** [comment]
> This does work (in theory) but what about the implications of SEO? Search engines generally prefer full URLs

You cannot have SEO if you never render the content to HTML, right? As for Google (which is able to crawl pages using client-side JavaScript navigation), that's a good question. @itsezc Up for asking AI about it? I guess there must be some content out there documenting this. I think Vike can support both approaches (see below), but it'd be nice to dig into this question so that we'll be able to add a note about it in Vike's docs.

> Vike option [...] to have `/product/42`

I guess this would do the trick:
 - Vike pre-renders `/product/+Page.js` with `pageContext.routeParams.id === null`. Thus the user's `+Page.js` must render some kind of empty/loading shell upon pre-rendering.
 - On the client-side, Vike renders `/product/42` in two steps:
   1. If [hydrationCanBeAborted](https://vike.dev/hydrationCanBeAborted) is `false` (Vue) then the page is hydrated with `pageContext.routeParams.id === null` (to avoid hydration mismatch), otherwise (React) the page isn't hydrated at all.
   2. The page is (re-)rendered with `pageContext.routeParams.id === '42'`.

So basically what [a comment above is describing](https://github.com/vikejs/vike/issues/1476#issuecomment-2636383336).

**Comment by micah-redwood** [comment]
I don't know if there's going to be an elegant solution to this that integrates well with static site hosts since AFAIK none of the popular ones have a way of defining dynamic routes in your HTML folder structure itself. A quick survey:

- Cloudflare Pages will render the _closest 404.html file_ to the requested route, meaning that as long as there is only one dynamic path segment and it's at the end of the route, you could generate a `404.html` to handle the request:
https://developers.cloudflare.com/pages/configuration/serving-pages/#not-found-behavior
- Netlify / Vercel would require setting up custom rewrite rules for every dynamic path segment:
https://vercel.com/guides/custom-404-page#what-if-i-need-to-name-my-404-file-something-different
https://docs.netlify.com/routing/redirects/rewrites-proxies/#shadowing
- All 3 providers (and presumably others) would let you use workers/snippets/functions to achieve this, but of course that comes with cost at scale.

I suppose the good news is that dynamic path segments are typically an edge case. Setting up your CDN to serve the right HTML file is a pain, but it's probably something you'll be doing for only a few routes, so it feels doable.

I think Vike's responsibility here is really just to output a consistently named file that will work well if the user goes to the trouble of getting their CDN to serve it. In the absence of a standard, maybe just generating the most predictable folder structure would make most sense:
`/product/@id/+Page.js` -> `/product/@id/index.html`

Then it's up to the user to rename that to `/product/404.html` for Cloudflare or setup rewrites for Vercel/Netlify, etc. To make things even easier, maybe Vike could generate a dynamic-routes.json file where it enumerates all dynamic routes so that it's easier for the user to generate the appropriate configs in a post-build script. Just thinking out loud.

One thing I will push back on: there was mention of using `/product/+Page.js` to render a shell that would then be reused for `/product/@id/`... this seems like a bad idea since it's conflating `/product` and `product/@id`. In this case it might be fine, but let's say I have `/products/` and `/products/@id/`. These are totally different pages (a list of products vs a specific product) with different layouts. So conflating the two seems like a real footgun.

In the interim I think I'll just use URL params as you suggested, something along the lines of `/products/product/?id=123` where I can customize `/products/product/+Page.js` and render my shell there.

**Comment by brillout** [comment]
> One thing I will push back on: there was mention of using `/product/+Page.js` to render a shell that would then be reused for `/product/@id/`... this seems like a bad idea since it's conflating `/product` and `product/@id`.

I agree and I think it should be a new setting `prerender.clientOnlyRouteParameter` while explaining that `pageContext.routeParams.id === null` must be handled when using that setting. Also a new section `https://vike.dev/pre-rendering#route-parameters` explaining the fundemental issue and the workarounds.

> Then it's up to the user to rename that to `/product/404.html` for Cloudflare or setup rewrites for Vercel/Netlify, etc. To make things even easier, maybe Vike could generate a dynamic-routes.json file where it enumerates all dynamic routes so that it's easier for the user to generate the appropriate configs in a post-build script. Just thinking out loud.

Yea, that's a good question. Long term the goal is to have `vike-*` extensions that automatically take care of that. In the meantime the user will be able to access `globalContext.prerenderContext.output.filter(o => o.isRouteParameterEmptyShell).map(o => o.filePath)` to programmatically handle this. Or, maybe even easier, the user could pass a function `prerender.clientOnlyRouteParameter: (o) => genUrlRewrite(o.route)`. It would be intereseting to see how Nuxt handle that: do they simply document the need to redirect URLs or do they have some kind of integretion packages?

> In the interim I think I'll just use URL params as you suggested, something along the lines of `/products/product/?id=123` where I can customize `/products/product/+Page.js` and render my shell there.

:+1: I didn't test it myself yet, but AFAICT it should work. Let me know if you run into any issues.

Contribution welcome to publish an example using that technique, if someone is up for it.

**Comment by brillout** [comment]
Thanks Micah and, indeed, React Router's design makes a lot of sense.

- `ssr: false` globally (without `prerender: true`) => a single `.html` file is generated `dist/client/__spa-fallback.html`
- `prerender: true` => one `__spa-fallback.html` per parameterized route
- The content of `__spa-fallback.html` is [`+Loading`](https://vike.dev/Loading)

I'm realizing that rendering `+Loading` is quite a good default (and better than what I proposed in my previous content).

As for configuring static hosts to redirect URLs to these `__spa-fallback.html`, we're in talks with deployment providers and there is an opportunity to make `__spa-fallback.html` a standard. So that we get end-to-end zero-config. (But it's going to take a little while until all deployment providers adopt it.)

**Comment by brillout** [comment]
Bumping this from `high-prio 💫` to `highest-prio 🔥` — we'll implement it as soon as we're done with our backlog (ETA 2026 Q1).

**Comment by brillout** [comment]
(We're still busy with higher priorities, but this is very much still on the radar.)

**Comment by micahjon** [comment]
Recently ran into this at work, where I needed to have dynamic top-level pages for each of our suppliers in the supplier portal, e.g. `/bmw`, `/panasonic` etc, and building it with Vike on Cloudflare Pages was easier than I expected:

1. Happy path (supplier slug known at build time) -> just pre-render a dynamic top-level route (`/@supplier/+Page.tsx`) and use `onBeforePrerenderStart` to grab all the necessary slugs. _Vike's already smart enough to prioritize hard-coded pages over dynamic ones so adding this top-level dynamic route segment didn't clobber any of my existing URLs and let me override some supplier-specific pages w/out jumping through any hoops_
2. Runtime 404 fallback (supplier slug added *after* build) -> use a smart 404 page that first checks to see if the path matches a supplier slug (does an API call) before actually displaying either the supplier page -or- the 404 page.

The fallback doesn't change the page's response status (still a 404, not a 200) and has an unavoidable loading state, but it gets the job done until the next re-deploy. Just wanted to share this pattern in case others would find it helpful. Obviously not ideal in scenarios where SEO matters a lot--in my case it wasn't important.
